### PR TITLE
Fix some problems with active failover 

### PR DESCRIPTION
**Pull request by graetzer**

### CODE REVIEW

File: Documentation/Books/Manual/Administration/Replication/Asynchronous/Limitations.md [hunk 1]
```diff
@@ -4,39 +4,29 @@ Replication Limitations
...
   operations on a slave. ArangoDB currently does not prevent users from carrying out
   their own write operations on slaves, though this might lead to undefined behavior
   and the replication applier stopping.
-* when a replication slave asks a master for log events, the replication master will 
-  return all write operations for user-defined collections, but it will exclude write
-  operations for certain system collections. The following collections are excluded
-  intentionally from replication: *_apps*, *_trx*, *_replication*, *_configuration*,
-  *_jobs*, *_queues*, *_sessions*, *_foxxlog* and all statistics collections.
+* The following collections are excluded
+  intentionally from replication: *_trx*, *_replication*, *_configuration*,
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/Asynchronous/Limitations.md]: Not true anymore for: 
* _trx
* _replication
* _fishbowl

File: Documentation/Books/Manual/Administration/Replication/Asynchronous/Limitations.md [hunk 2]
```diff
@@ -4,39 +4,29 @@ Replication Limitations
...
-  possible to have a slave apply operations from multiple masters into the same target
-  database.
-* replication is set up on a per-database level. When using ArangoDB with multiple
-  databases, replication must be configured individually for each database.
-* the replication applier is single-threaded, but write operations on the master may
-  be executed in parallel if they affect different collections. Thus the replication
-  applier might not be able to catch up with a very powerful and loaded master.
+* It is currently not possible to have a follower apply operations from multiple Leaders into the
+  same target database.
+* the replication applier is single-threaded, but write operations may
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/Asynchronous/Limitations.md]: The replication applier (capital T)

File: Documentation/Books/Manual/Administration/Replication/Asynchronous/README.md [hunk 3]
```diff
@@ -1,66 +1,70 @@
 Asynchronous replication
 ========================
 
-Asynchronous replication works by logging every data modification on a *master* and replaying these events on a number of *slaves*.
+Asynchronous replication works by logging every data modification on a *Leader* server and replaying these events on a
+onder or more *Follower* servers.
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/Asynchronous/README.md]: "onder" => "one"

File: Documentation/Books/Manual/Administration/Replication/Asynchronous/README.md [hunk 4]
```diff
@@ -1,66 +1,70 @@
...
 applier must be set to *true*. 
 
 #### Replication overhead
 
-As the master servers are logging any write operation in the write-ahead-log anyway replication doesn't cause any extra overhead on the master. However it will of course cause some overhead for the master to serve incoming read requests of the slaves. Returning the requested data is however a trivial task for the master and should not result in a notable performance degration in production.
\ No newline at end of file
+As the Leader servers are logging any write operation in the write-ahead-log anyway replication does not cause
+any unreasonable overhead on the Leader node. However it will of course cause some overhead to serve
+incoming read requests, namely it will use up some disk and network bandwidth. Returning the requested data is however
+fairly a trivial task and should not result in a notable performance degration in production.
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/Asynchronous/README.md]: degradation

File: Documentation/Books/Manual/Administration/Replication/Asynchronous/ServerLevelSetup.md [hunk 5]
```diff
@@ -1,36 +1,37 @@
...
+the Leader
 
-Setting up a working master-slave replication requires two ArangoDB instances:
-* **master**: this is the instance that all data-modification operations should be directed to
-* **slave**: on this instance, we'll start a replication applier, and this will fetch data from the 
-  master database's write-ahead log and apply its operations locally
+Setting up a working classical master-slave replication requires always at least two ArangoDB instances:
+* **Leader**: this is the instance that all data-modification operations should be directed to
+* **Follower**: on this instance, we'll start a replication applier, and this will fetch data from the
+  Leader node's write-ahead log and apply its operations locally (there can be more than one Follower per Leader)
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/Asynchronous/ServerLevelSetup.md]: Do we want to advertise this already?

File: Documentation/Books/Manual/Administration/Replication/Asynchronous/SyncingCollections.md [hunk 6]
```diff
@@ -1,27 +1,27 @@
...
+In order to synchronize data for a single collection from a Leader to a slave instance, there
 is the *syncCollection* function:
 
-It will fetch all documents of the specified collection from the master database and store 
-them in the local instance. After the synchronization, the collection data on the slave will be
-identical to the data on the master, provided no further data changes happen on the master. 
-Any data changes that are performed on the master after the synchronization was started will
+It will fetch all documents of the specified collection from the Leader server's database and store
+them in the local instance. After the synchronization, the collection data on the Folloer will be
+identical to the data on the Leader, provided no further data changes happens.
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/Asynchronous/SyncingCollections.md]: happens => happen

File: Documentation/Books/Manual/Administration/Replication/README.md [hunk 7]
```diff
@@ -5,21 +5,49 @@ Replication allows you to *replicate* data onto another machine. It
 forms the base of all disaster recovery and failover features ArangoDB
 offers. 
 
-ArangoDB offers asynchronous and synchronous replication which both
-have their pros and cons. Both modes may and should be combined in a
-real world scenario and be applied in the usecase where they excel
-most. 
+ArangoDB offers **asynchronous** and **synchronous** replication,
+depending on which type of arangodb deployment you are using.
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/README.md]: "arangodb" => "ArangoDB"

File: Documentation/Books/Manual/Administration/Replication/README.md [hunk 8]
```diff
@@ -5,21 +5,49 @@ Replication allows you to *replicate* data onto another machine. It
...
-sections. 
+Since ArangoDB 3.2 the *synchronous replication* replication is the *only* replication
+type used in a cluster whereas the asynchronous replication is only available between
+single-server nodes. Future versions of ArangoDB may reintroduce asynchronous
+replication for the cluster.
+
+### Asynchronous replication
+
+In ArangoDB any write, delete or update operation will be logged to the write-ahead
+log (WAL), before beeing acknowledged to the client. The write ahead ensures
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/README.md]: "beeing" => "being"
"to the client" => "by the client"

File: Documentation/Books/Manual/Administration/Replication/README.md [hunk 9]
```diff
@@ -5,21 +5,49 @@ Replication allows you to *replicate* data onto another machine. It
...
+that in the occurance of an unscheduled server shutdown, we are still
+able to restore already commited transactions.
+
+The asynchronous replication is used as the foundation of most replication operations in
+arangodb. Effecitvely you as a potential end-user will **only** ever encouter it directly
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/README.md]: "arangodb" => "ArangoDB"

Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/README.md]: "potential end-user" => "end-user"

Comment (jsteemann) [reply Documentation/Books/Manual/Administration/Replication/README.md]: "encouter" => "encounter"

File: Documentation/Books/Manual/Administration/Replication/README.md [hunk 10]
```diff
@@ -5,21 +5,49 @@ Replication allows you to *replicate* data onto another machine. It
...
+**on single-instance**nodes.
+
+The WAL is also useful for other purposes than data-integrety, it also allows
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/README.md]: "integrety" => "integrity"

File: Documentation/Books/Manual/Administration/Replication/README.md [hunk 11]
```diff
@@ -5,21 +5,49 @@ Replication allows you to *replicate* data onto another machine. It
...
+other replication clients to follow along with all changes made on a specific *leading* server.
+These async-replication clients are called *followers*, which will periodically poll the *leader*
+for changes to it's data. The clients will apply all modification operations in the same order
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/README.md]: "it's" => "its"

File: Documentation/Books/Manual/Administration/Replication/README.md [hunk 12]
```diff
@@ -5,21 +5,49 @@ Replication allows you to *replicate* data onto another machine. It
...
-used for mission critical data which must be accessible at all
-times. Synchronous replication generally stores a copy of a shard's
-data on another db server and keeps it in sync. Essentially, when storing
-data after enabling synchronous replication the cluster will wait for
+Synchronous replication is only available in the context of an ArangodB **cluster setup**.
+In the case of synchronous replication we are always talking about replicating **shards** of a collection
+(a shard is a fragment of a collection, stored on one DBServer node). Synchronous replication guarantees that
+replicated copies of a *leader* shard are stored on another *follower* db server and kept completely **in sync**.
+
+By "in-sync" we mean that we guarantees *strong consistency* between replica shards: Your data is
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/README.md]: "in-sync" (here) vs. "in sync" (in line above).
We should write it consistently.

File: Documentation/Books/Manual/Administration/Replication/README.md [hunk 13]
```diff
@@ -5,21 +5,49 @@ Replication allows you to *replicate* data onto another machine. It
...
+available in an identical version on all "in-sync" followers. This is in contrast to *eventual consistency*
+where a follower server might contain slightly older data than the leader server. Synchronously replicated
+shards are always identical to their leader shard and can take over in case of a server outage.
+
+Synchronous replication can even work reliably for for mission critical data which must be accessible at all
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/README.md]: "for for" => "for"

File: Documentation/Books/Manual/Administration/Replication/README.md [hunk 14]
```diff
@@ -5,21 +5,49 @@ Replication allows you to *replicate* data onto another machine. It
...
+times. . Essentially, when storing data after enabling synchronous replication the cluster will wait for
```
Comment (jsteemann) [inline Documentation/Books/Manual/Administration/Replication/README.md]: ". ." => ". "

File: arangod/Cluster/ClusterInfo.cpp [hunk 15]
```diff
@@ -1196,9 +1196,9 @@ int ClusterInfo::createCollectionCoordinator(std::string const& databaseName,
                     p.value, "error", false)) {
               tmpError += " shardID:" + p.key.copyString() + ":";
               tmpError += arangodb::basics::VelocyPackHelper::getStringValue(
-                  p.value, "errorMessage", "");
-              if (p.value.hasKey("errorNum")) {
-                VPackSlice const errorNum = p.value.get("errorNum");
+                  p.value, "error", "");
```
Comment (jsteemann) [inline arangod/Cluster/ClusterInfo.cpp]: should be "errorMessage"!!

File: arangod/Cluster/HeartbeatThread.cpp [hunk 16]
```diff
@@ -614,34 +593,34 @@ void HeartbeatThread::runSingleServer() {
...
         config._autoResync = true;
-        config._autoResyncRetries = 3;
+        config._autoResyncRetries = 2;
         // TODO: how do we initially configure the applier
         
-        // start initial synchronization
+        LOG_TOPIC(INFO, Logger::HEARTBEAT) << "start initial sync from leader";
         TRI_ASSERT(!config._skipCreateDrop);
-        GlobalInitialSyncer syncer(config);
+        /*GlobalInitialSyncer syncer(config);
```
Comment (jsteemann) [inline arangod/Cluster/HeartbeatThread.cpp]: can we remove this entire section if it is unnecessary?

File: arangod/Cluster/HeartbeatThread.cpp [hunk 17]
```diff
@@ -614,34 +593,34 @@ void HeartbeatThread::runSingleServer() {
...
+        TRI_voc_tick_t lastLogTick = syncer.getLastLogTick();*/
 
         // forget about any existing replication applier configuration
         applier->forget();
-        
         applier->reconfigure(config);
-        LOG_TOPIC(DEBUG, Logger::HEARTBEAT) << "now starting the applier from initial tick " << lastLogTick;
-        applier->start(lastLogTick, true, barrierId);
+        applier->startReplication();
+        //LOG_TOPIC(DEBUG, Logger::HEARTBEAT) << "now starting the applier from initial tick " << lastLogTick;
```
Comment (jsteemann) [inline arangod/Cluster/HeartbeatThread.cpp]: remove as well

File: arangod/Replication/ReplicationApplierState.h [hunk 18]
```diff
@@ -36,10 +36,11 @@ namespace arangodb {
 
 /// @brief state information about replication application
 struct ReplicationApplierState {
-  enum class ActivityState {
-    INACTIVE,
-    RUNNING,
-    SHUTTING_DOWN
+  enum class ActivityPhase {
+    INACTIVE, /// sleeping
+    INITAL, /// running inital syncer
```
Comment (jsteemann) [inline arangod/Replication/ReplicationApplierState.h]: should ne INITIAL (one I missing)

File: arangod/Replication/TailingSyncer.cpp [hunk 19]
```diff
@@ -1062,8 +1062,11 @@ void TailingSyncer::getLocalState() {
     return;
   }
  
+  // a _masterInfo._serverId value of 0 may occur if no proper connection could be
+  // established to the master initially
   if (_masterInfo._serverId != _applier->_state._serverId &&
-      _applier->_state._serverId != 0) {
+      _applier->_state._serverId != 0 && _masterInfo._serverId != 0) {
+    LOG_TOPIC(ERR, Logger::REPLICATION) << "OUR MASTERINFO: " << (void*) &_masterInfo;
```
Comment (jsteemann) [inline arangod/Replication/TailingSyncer.cpp]: oops, we must remove this!

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 20]
```diff
@@ -494,17 +494,17 @@ void RestReplicationHandler::handleCommandMakeSlave() {
   // will throw if invalid
   configuration.validate();
 
-  std::unique_ptr<InitialSyncer> syncer;
+  /*std::unique_ptr<InitialSyncer> syncer;
```
Comment (jsteemann) [inline arangod/RestHandler/RestReplicationHandler.cpp]: remove?

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 21]
```diff
@@ -494,17 +494,17 @@ void RestReplicationHandler::handleCommandMakeSlave() {
...
   } else {
     syncer.reset(new DatabaseInitialSyncer(_vocbase, configuration));
-  }
+  }*/
 
   // forget about any existing replication applier configuration
   applier->forget();
 
-  // start initial synchronization
+  /*// start initial synchronization
```
Comment (jsteemann) [inline arangod/RestHandler/RestReplicationHandler.cpp]: remove?

File: Documentation/Books/Manual/Administration/Replication/Asynchronous/ServerLevelSetup.md (hunk 5 above)
Comment (graetzer) [reply Documentation/Books/Manual/Administration/Replication/Asynchronous/ServerLevelSetup.md]: yes, this is not yet the Active-Passive Setup

**Pull request merged**